serviceregistration/kubernetes/retry_handler.go

```diff
@@ -0,0 +1,112 @@
+package kubernetes
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/serviceregistration/kubernetes/client"
+)
+
+// How often to retry sending a state update if it fails.
+var retryFreq = 5 * time.Second
+
+// retryHandler executes retries.
+// It is thread-safe.
+type retryHandler struct {
+	// These don't need a mutex because they're never mutated.
+	logger             hclog.Logger
+	namespace, podName string
+	client             *client.Client
+
+	// This gets mutated on multiple threads.
+	patchesToRetry     []*client.Patch
+	patchesToRetryLock sync.Mutex
+}
+
+// Run runs at an interval, checking if anything has failed and if so,
+// attempting to send them again.
+func (r *retryHandler) Run(shutdownCh <-chan struct{}, wait *sync.WaitGroup) {
+	// Make sure Vault will give us time to finish up here.
+	wait.Add(1)
+	defer wait.Done()
+
+	retry := time.NewTicker(retryFreq)
+	for {
+		select {
+		case <-shutdownCh:
+			return
+		case <-retry.C:
+			r.retry()
+		}
+	}
+}
+
+// Add adds a patch to be retried until it's either completed without
+// error, or no longer needed.
+func (r *retryHandler) Add(patch *client.Patch) error {
+	r.patchesToRetryLock.Lock()
+	defer r.patchesToRetryLock.Unlock()
+
+	// - If the patch is a dupe, don't add it.
+	// - If the patch reverts another, remove them both.
+	//     For example, perhaps we were already retrying "active = true",
+	//     but this new patch tells us "active = false" again.
+	// - Otherwise, this is a new, unique patch, so add this patch to retries.
+	for i := 0; i < len(r.patchesToRetry); i++ {
+		prevPatch := r.patchesToRetry[i]
+		if patch.Path != prevPatch.Path {
+			continue
+		}
+		if patch.Operation != prevPatch.Operation {
+			continue
+		}
+		// These patches are operating upon the same thing.
+		// Let's look at what they're trying to do to determine
+		// the right action to take with the incoming patch.
+		patchValStr, ok := patch.Value.(string)
+		if !ok {
+			return fmt.Errorf("all patches must have bool values but received %+x", patch)
```

Comment:
Is this restriction, at this level, warranted?

Reply:
It makes the code easier to reason about. The retryHandler isn't used for anything else. Arguably, it could not be broken out as its own object. I tried coding it up that way but I thought it made the service registration code harder to reason about, and it made the code harder to unit test. They are indeed not 100% decoupled.

Reply:
Does this conflict with VaultVersion?

Reply:
Vault version is static and never changes, and so we don't need to incorporate it into the retryHandler's logic. There's likewise no NotivyVaultVersionChange method on the service registration interface. We just get it passed in at the beginning the the Factory method then never hear of it again.